### PR TITLE
fix: reserve submitted redemption burn receipts

### PR DESCRIPTION
**Pull request by JuaniRios**
## Motivation

A production redemption retry exposed a race in receipt burn planning: one redemption can submit a Fireblocks burn for a receipt, then remain pending (or fail ambiguously) during confirmation, while a second redemption plans against the same receipt balance before the first burn is reflected in receipt inventory. That produces a later on-chain `ERC1155InsufficientBalance` revert even though the second redemption appeared to have enough inventory at planning time.

The fix is to treat a submitted burn as *committed* inventory the moment it is submitted — and to resolve that commitment correctly across the full lifecycle: consume it when the burn confirms, release it only when the burn definitively did not consume any shares, and never release it while the outcome is ambiguous (the transaction may still land).

Closes [RAI-649](https://linear.app/makeitrain/issue/RAI-649/same-receipt-burn-race-concurrent-redemptions-reuse-submitted-burn).

Related follow-up: [RAI-646](https://linear.app/makeitrain/issue/RAI-646/separate-ambiguous-fireblocks-burn-confirmation-from-definitive-burn).

## Solution

**Two-quantity inventory model.** Each `ReceiptMetadata` now tracks a `balance` (the on-chain ERC-1155 balance the bot believes it holds) plus a `reserved` map of shares committed to submitted-but-unconfirmed burns, keyed by redemption. Available inventory for planning is `balance − sum(reserved)`. Keeping the mirror and the reservations separate is what lets reconciliation compare against the true on-chain balance without clobbering a pending reservation. `Shares` gains `ZERO` and `Ord`.

**Reservation lifecycle** (new `ReceiptInventoryCommand`/`Event` variants):

- `ReserveBurn` / `BurnReserved` — **atomic clear-and-reserve**, issued *before* submitting the burn to the signing backend. It drops any prior reservation the redemption held (e.g. a retry that planned different receipts) and installs the new one in one serialized transaction, validated against availability that excludes the redemption's own prior reservation. The prior reservation is never returned to global availability between clear and reserve, so a concurrent *different* redemption that already committed the same balance fails to reserve and never submits an unbacked burn.
- `SettleBurn` / `BurnSettled` — consume the reservation after the burn confirms on-chain: clear it and reduce the mirror balance by the reserved amount. Idempotent; emits `Depleted` for any receipt it empties.
- `ReleaseBurn` / `BurnReleased` — clear the redemption's reservation, restoring availability without touching the mirror (the burn consumed nothing). No-op when no reservation is held.

**Planning excludes the planner's own reservation.** `for_burn` now takes the redemption id and plans against `available_excluding(redemption)`, so a retry re-plans its full burn rather than being blocked by its own prior reservation; the subsequent atomic `ReserveBurn` then replaces it.

**`BurnManager` wiring** (`src/redemption/burn_manager.rs`):

- Reserve planned burns *before* submission via `reserve_with_conflict_retry` (bounded retry on optimistic-concurrency conflict, since concurrent redemptions contend on the single per-vault inventory aggregate). On reserve failure, record the burn failure and abort rather than submit an unbacked burn.
- Settle on confirmation (live path and `BurnSubmitted`/`BurnFailed` recovery paths).
- Release only on a **definitive** failure, gated by the new `should_release_reserved_burn` helper: a terminal Fireblocks status or an EVM revert. Ambiguous/pending statuses, missing-tx-id, and other non-vault errors **retain** the reservation so a concurrent redemption can't reuse the balance and double-submit.

**Definitive on-chain revert detection** (new `VaultError::Reverted`). Both backends now treat a mined-but-reverted burn (`receipt.status() == false`) as a definitive no-consumption failure, distinct from an anomalous missing-`Withdraw` parse error: the local backend catches it synchronously at submit (`src/vault/service.rs`), the Fireblocks backend at confirm time (`src/fireblocks/vault_service.rs`). `is_still_pending` is re-exported from the fireblocks module so the burn manager can distinguish ambiguous-pending from definitive failure.

**Reconciliation respects pending reservations.** `ReconcileBalance` now treats any on-chain balance within the `[available, balance]` band as "no external change" — while a submitted burn is pending, on-chain legitimately sits there (not yet landed at `balance`, dropping toward `available` once it lands). A reserved receipt drained to zero on-chain is preserved (mirror set to zero) rather than removed, so its reservation can still resolve; depletion is deferred until the last reservation clears. Settlement, not reconciliation, consumes a reservation.

**Startup reservation recovery** (`recover_stuck_reservations`, run after redemption recovery). For each vault it resolves dangling reservations against the redemption's terminal state: a `Completed` redemption is **settled** (confirmed but settlement was missed, e.g. a crash in the confirm→settle window); `Failed`/`Closed`/unknown are **left in place**, because a surviving reservation there is from an *ambiguous* failure whose burn may still have landed — releasing would over-credit inventory and risk a duplicate burn.

**Tests.** Extensive aggregate coverage for reserve/settle/release (idempotency, atomic replace-on-retry, second-redemption rejection, reconcile band / zero-drain / pending-reservation preservation), `BurnManager` coverage for retain-vs-release across submit/confirm reverts and Fireblocks statuses, and `recover_stuck_reservations` (settles completed, leaves failed/unknown). New mock behaviors `new_submit_revert`/`new_confirm_revert`. Burn-planner one-gwei boundary tests added.

**Docs.** `SPEC.md` updated with the reservation model, the reconcile band, and the startup recovery semantics.

Stacked on #162.

## Checks

By submitting this for review, I'm confirming I've done the following:

- [x] added comprehensive test coverage for any changes in logic
- [x] made this PR as small as possible
- [x] linked any relevant issues or PRs

Tested with:

- `direnv exec . cargo test -q receipt_inventory`
- `direnv exec . cargo test -q burn_manager`
- `direnv exec . cargo test -q`

## Summary by CodeRabbit

* **New Features**
  * Added burn reservation and release lifecycle management to prevent concurrent redemptions from reusing committed balances
  * Introduced startup recovery mechanism for stuck reservations based on terminal redemption states
  * Enhanced on-chain transaction monitoring with explicit revert detection

* **Bug Fixes**
  * Improved reconciliation behavior to preserve pending burn reservations
  * Better error handling for insufficient per-receipt balances

* **Tests**
  * Extended test coverage for burn reservation edge cases, settlement behavior, and revert scenarios

* **Documentation**
  * Updated specification for burn-reservation timing and inventory aggregate lifecycle

* **Chores**
  * Updated .gitignore to exclude internal directories

[Review Change Stack](https://app.coderabbit.ai/change-stack/ST0x-Technology/st0x.issuance/pull/163?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

**Comment by JuaniRios** [comment]
* **#164** <a href="https://app.graphite.com/github/pr/ST0x-Technology/st0x.issuance/164?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#163** <a href="https://app.graphite.com/github/pr/ST0x-Technology/st0x.issuance/163?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.com/github/pr/ST0x-Technology/st0x.issuance/163?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#162** <a href="https://app.graphite.com/github/pr/ST0x-Technology/st0x.issuance/162?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#161** <a href="https://app.graphite.com/github/pr/ST0x-Technology/st0x.issuance/161?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add the label <kbd>add-to-gt-merge-queue</kbd> to this PR to add it to the merge queue.

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.com/invite/github/ST0x-Technology?ref=merge-queue-instructions-in-stack-comment&prId=163).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: src/receipt_inventory/burn_tracking.rs [hunk 1]
```diff
@@ -743,6 +743,79 @@ mod tests {
...
+        let err = plan_burn(receipts, burn_amount, dust_amount)
+            .expect_err("Plan should fail before Fireblocks submission");
+
+        assert!(
+            matches!(err, BurnTrackingError::InsufficientBalance { required, available }
+                if required == burn_amount
+                    && available == Shares::new(uint!(10079999999000000000_U256))),
+            "Expected one-wei InsufficientBalance error, got {err:?}"
+        );
+    }
```
[the remark this answers is not in the document]
Comment (JuaniRios) [reply src/receipt_inventory/burn_tracking.rs:817]: Thanks, but this no longer applies. These tests are named `test_plan_spans_two_receipts_when_first_is_short_by_one_gwei` and `test_plan_fails_when_single_receipt_short_by_one_gwei` — "one gwei", not "one wei". The 1,000,000,000-wei (= 1 gwei) gap matches the names, so there's no discrepancy to fix. The boundary we intend to cover here is the one-gwei case, which the values already exercise.

**Review by findolor — APPROVED** [review]

File: src/receipt_inventory/mod.rs [hunk 2]
```diff
@@ -81,10 +83,14 @@ impl std::fmt::Display for ReceiptId {
 ///
 /// Shares represent ownership in the vault and are minted/burned
 /// proportionally to deposited/withdrawn assets.
-#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
+#[derive(
+    Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Serialize, Deserialize,
+)]
 pub(crate) struct Shares(U256);
 
 impl Shares {
```
Comment (0xgleb) [inline src/receipt_inventory/mod.rs:91]: i guess we need to consider pulling out st0x-finance crate from st0x.liquidity repo into a reusable thing too at some point

File: .claude/scheduled_tasks.lock
Comment (0xgleb) [inline .claude/scheduled_tasks.lock:1]: looks user-specific, no?

File: src/fireblocks/mod.rs [hunk 3]
```diff
@@ -13,6 +13,7 @@ pub(crate) use config::{
 };
 pub(crate) use vault_service::{
     FireblocksVaultError, FireblocksVaultService, fetch_vault_address,
+    is_still_pending as fireblocks_transaction_is_still_pending,
```
Comment (0xgleb) [inline src/fireblocks/mod.rs]: just qualifying `vault_service::is_still_pending` imo is both more readable, less verbose, and less ambiguous than import aliases

File: src/redemption/burn_manager.rs [hunk 4]
```diff
@@ -168,6 +172,105 @@ where
...
+            match self.receipt_service.reserved_redemptions(*vault).await {
+                Ok(redemptions) => {
+                    stuck
+                        .extend(redemptions.into_iter().map(|id| (*vault, id)));
+                }
+                Err(err) => {
+                    warn!(target: "redemption", vault = %vault, error = %err,
+                        "Failed to list reserved redemptions during reservation recovery"
+                    );
+                }
```
Comment (0xgleb) [inline src/redemption/burn_manager.rs]: nit: consider

```suggestion
                Err(error) => {
                    warn!(target: "redemption", %vault, %error,
                        "Failed to list reserved redemptions during reservation recovery"
                    );
                }
```

File: src/receipt_inventory/mod.rs (hunk 2 above)
Comment (JuaniRios) [reply src/receipt_inventory/mod.rs:91]: Tracked here: https://linear.app/makeitrain/issue/RAI-669/reuse-existing-st0x-finance-crate-from-issuance

File: .claude/scheduled_tasks.lock
Comment (JuaniRios) [reply .claude/scheduled_tasks.lock:1]: Fixed in 1850784.

File: src/fireblocks/mod.rs (hunk 3 above)
Comment (JuaniRios) [reply src/fireblocks/mod.rs]: Fixed in 1850784.

File: src/redemption/burn_manager.rs (hunk 4 above)
Comment (JuaniRios) [reply src/redemption/burn_manager.rs]: Fixed in 1850784.

**Review by 0xgleb — APPROVED** [review]
